File: server/app/routes/auth_routes.py
from flask import Blueprint, request, jsonify
from app.config.firebaseSetup import db
import firebase_admin
from firebase_admin import auth, credentials
from dotenv import load_dotenv
import requests
import os

load_dotenv()

auth_bp = Blueprint('auth', __name__, url_prefix='/auth')


# # Load Firebase credentials
# cred = credentials.Certificate(os.getenv("FIREBASE_KEY_PATH"))
# if not firebase_admin._apps:
#     firebase_admin.initialize_app(cred)

GOOGLE_CLIENT_ID = os.getenv("GOOGLE_CLIENT_ID")  # Ensure this is in your .env file

def verify_google_token(token):
    """
    Verify Google OAuth Token with Google API.
    """
    google_verify_url = f"https://oauth2.googleapis.com/tokeninfo?id_token={token}"
    response = requests.get(google_verify_url)
    if response.status_code == 200:
        return response.json()
    return None

@auth_bp.route('/google-login', methods=['POST'])
def google_login():
    """
    Log in or sign up users using Google OAuth.
    """
    data = request.get_json()
    id_token = data.get("id_token")

    if not id_token:
        return jsonify({"message": "ID token is required"}), 400

    # Verify Google ID token
    google_user = verify_google_token(id_token)

    if not google_user:
        return jsonify({"message": "Invalid Google token"}), 401

    # Extract user info
    email = google_user.get("email")
    name = google_user.get("name")
    uid = google_user.get("sub")  # Google's unique user ID

    try:
        # Check if the user already exists in Firebase Auth
        firebase_user = auth.get_user_by_email(email)
    except firebase_admin.auth.UserNotFoundError:
        # If user does not exist, create one in Firebase
        firebase_user = auth.create_user(uid=uid, email=email, display_name=name)

    # Generate a Firebase Custom Token (Optional)
    custom_token = auth.create_custom_token(firebase_user.uid)

    return jsonify({
        "message": "Login successful",
        "uid": firebase_user.uid,
        "email": email,
        "token": custom_token.decode("utf-8")  # Convert bytes to string
    }), 200



@auth_bp.route('/signup', methods=['POST'])
def signup():
    # Mock signup. (Need to set up Firebase Auth)
    data = request.json
    email = data.get('email')
    password = data.get('password')
    try:
        user = auth.create_user(email=email, password=password)
        return jsonify({"message": "User created successfully", "uid": user.uid}), 201
    except Exception as e:
        return jsonify({"message": str(e)}), 400
    
@auth_bp.route('/login', methods=['POST'])
def login():
    # Mock login
    data = request.get_json()
    email = data.get('email')
    password = data.get('password')
    try:
        user = auth.get_user_by_email(email)
        return jsonify({"message": "User logged in successfully", "uid": user.uid}), 200
    except Exception as e:
        return jsonify({"message": str(e)}), 400

File: server/routes/authRoutes.js
import express from 'express';

import { getAuth } from 'firebase-admin/auth'; // Fix for 'auth' is not defined
import { signInWithCredential, GoogleAuthProvider } from 'firebase/auth'; // Fix missing imports
import { registerWithEmail, loginWithEmail, resetPassword } from '../services/authService.js';
import { googleClient } from '../utils/firebaseAdmin.js'; // Ensure googleClient is defined in firebase.js

const router = express.Router();
const auth = getAuth(); // Firebase Admin SDK Auth Instance

// ✅ Register User
router.post('/register', async (req, res) => {
    try {
        const { email, password, name } = req.body;
        if (!email || !password || !name) return res.status(400).json({ error: 'Missing fields' });

        const user = await registerWithEmail(email, password, name);
        res.json({ success: true, user });
    } catch (error) {
        res.status(400).json({ error: error.message });
    }
});

// ✅ Login User
router.post('/login', async (req, res) => {
    try {
        const { email } = req.body;
        if (!email) return res.status(400).json({ error: 'Missing email' });v 

        const user = await loginWithEmail(email);
        res.json({ success: true, user });
        console.log('SUccess User:', user);
    } catch (error) {
        res.status(400).json({ error: error.message });
    }
});

// ✅ Reset Password
router.post('/reset-password', async (req, res) => {
    try {
        const { email } = req.body;
        if (!email) return res.status(400).json({ error: 'Missing email' });

        const response = await resetPassword(email);
        res.json(response);
    } catch (error) {
        res.status(400).json({ error: error.message });
    }
});

// ✅ Google Sign-In Route (Fixes missing imports & undefined variables)
router.post('/google-login', async (req, res) => {
    try {
        const { id_token } = req.body;
        
        if (!id_token) {
            return res.status(400).json({ error: 'Missing ID token' });
        }

        // ✅ Verify Google token
        const ticket = await googleClient.verifyIdToken({
            idToken: id_token,
            audience: process.env.VITE_APP_GOOGLE_OAUTH_CLIENT_ID
        });

        const payload = ticket.getPayload();
        const credential = GoogleAuthProvider.credential(id_token);
        
        try {
            const userCredential = await signInWithCredential(auth, credential);
            res.status(200).json({
                success: true,
                user: {
                    uid: userCredential.user.uid,
                    email: payload.email,
                    name: payload.name,
                    picture: payload.picture
                }
            });
        } catch (firebaseError) {
            console.error('Firebase authentication error:', firebaseError);
            res.status(401).json({
                error: 'Firebase authentication failed',
                details: firebaseError.message
            });
        }
    } catch (error) {
        console.error('Google authentication error:', error);
        res.status(401).json({
            error: 'Authentication failed',
            details: error.message
        });
    }
});

export default router;
